internal/masterservice/timeticksync.go

```diff
@@ -0,0 +1,172 @@
+		chanStream:    make(map[string]msgstream.MsgStream),
+		sendChan:      make(chan map[typeutil.UniqueID]*internalpb.ChannelTimeTickMsg),
+		addChan:       addChan,
+		delChan:       delChan,
+	}
+
+	pnSessions := gm.GetSessions(proxyNodePrefix)
+	for _, pns := range pnSessions {
+		tss.proxyTimeTick[pns.ServerID] = nil
+	}
+	return &tss, nil
+}
+
+// sendToChannel send all channels' timetick to sendChan
+func (t *timetickSync) sendToChannel() {
+	for _, v := range t.proxyTimeTick {
+		if v == nil {
+			return
+		}
+	}
+	// clear proxyTimeTick and send a clone
+	ptt := make(map[typeutil.UniqueID]*internalpb.ChannelTimeTickMsg)
+	for k, v := range t.proxyTimeTick {
+		ptt[k] = v
+		t.proxyTimeTick[k] = nil
+	}
+	t.sendChan <- ptt
+}
+
+// UpdateTimeTick check msg validation and send it to local channel
+func (t *timetickSync) UpdateTimeTick(in *internalpb.ChannelTimeTickMsg) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	_, ok := t.proxyTimeTick[in.Base.SourceID]
+	if !ok {
+		return fmt.Errorf("Skip ChannelTimeTickMsg from un-recognized proxy node %d", in.Base.SourceID)
+	}
+	t.proxyTimeTick[in.Base.SourceID] = in
+	t.sendToChannel()
+	return nil
+}
+
+// StartWatch watch proxy node change and process all channels' timetick msg
+func (t *timetickSync) StartWatch() {
+	for {
+		select {
+		case <-t.ctx.Done():
+			log.Debug("timetickSync context done", zap.Error(t.ctx.Err()))
+			return
+		case pns := <-t.addChan:
+			t.lock.Lock()
+			t.proxyTimeTick[pns.ServerID] = nil
+			t.lock.Unlock()
+		case pns := <-t.delChan:
+			t.lock.Lock()
+			delete(t.proxyTimeTick, pns.ServerID)
+			t.sendToChannel()
+			t.lock.Unlock()
+		case ptt, ok := <-t.sendChan:
+			if !ok {
+				log.Debug("timetickSync sendChan closed", zap.Error(t.ctx.Err()))
+				return
+			}
+			// reduce each channel to get min timestamp
+			chanName2TimeTickMap := make(map[string]typeutil.Timestamp)
+			for _, cttMsg := range ptt {
+				chanNum := len(cttMsg.ChannelNames)
+				for i := 0; i < chanNum; i++ {
+					name := cttMsg.ChannelNames[i]
+					ts := cttMsg.Timestamps[i]
+					cts, ok := chanName2TimeTickMap[name]
+					if !ok || ts < cts {
+						chanName2TimeTickMap[name] = ts
+					}
+				}
+			}
+			// send timetick msg to msg stream
+			for chanName, chanTs := range chanName2TimeTickMap {
+				if err := t.SendChannelTimeTick(chanName, chanTs); err != nil {
+					log.Debug("SendChannelTimeTick fail", zap.Error(t.ctx.Err()))
+				}
+			}
+		}
+	}
+}
+
+// SendChannelTimeTick send each channel's min timetick to msg stream
+func (t *timetickSync) SendChannelTimeTick(chanName string, ts typeutil.Timestamp) error {
+	msgPack := msgstream.MsgPack{}
+	baseMsg := msgstream.BaseMsg{
+		BeginTimestamp: ts,
+		EndTimestamp:   ts,
+		HashValues:     []uint32{0},
+	}
+	timeTickResult := internalpb.TimeTickMsg{
+		Base: &commonpb.MsgBase{
+			MsgType:   commonpb.MsgType_TimeTick,
+			MsgID:     0,
+			Timestamp: ts,
+			SourceID:  0,
```

Comment:
`SourceID` should be `master service`'s node id

Reply:
done